### DISCUSSION TITLE
Add `ekg-edit-hook`

### DISCUSSION THREAD

**Discussion by qingshuizheng** [ahyatt/ekg#80] [Ideas]
### Senario: auto save without having to finalize
I don't finalize the ekg capture/edit buffer too often, just to keep them open for more info to come in. Very often I mistakenly kill an unsaved note buffer, and those are the moments I'm dying inside. :-(

So I have a workflow for auto-saving current ekg-capture/edit instance without having to finalize using `ekg-capture/edit-finalize`. The note auto saves every 1 sec using the goodies of the package `super-save` but with an advised `super-save-command`
```elisp
   (define-advice super-save-command (:override (&rest _) with-ekg-support)
     "Save the current buffer if needed."
     (if (and (or (bound-and-true-p ekg-edit-mode)
                  (bound-and-true-p ekg-capture-mode))
              (not (ekg--in-metadata-p)) ; don't auto save while in overlay, to avoid auto-saving useless tags
              (buffer-modified-p (current-buffer)))
         ;; ekg buffer
         (progn
           (ekg--update-from-metadata)
           (ekg--save-note-in-buffer)
           (message "Saved."))
       ;; other buffer
       (when (super-save-p) (save-buffer))))
```

If the note is saved via the above auto-save mechanism, the note will be tagged as "raw", so I have a clue to find them later on; If the note is finalized normally, the "raw" tag will be converted to "draft".

The workflow relies on the state of `buffer-modified-p` in capture/edit buffer. At the moment the capture/edit buffer's initial modified state is always `t`.  It should be set default to nil, otherwise it will trigger `super-save` to save when a note is captured or opened. 

I have tried it in my personal fork commit (bec068f) for one month or so,  which works great. But I cannot think of a good reason to ask for adding a `(set-buffer-modified-p nil)` to both `ekg-capture` and `ekg-edit`. Or maybe It's okay to add it?

### A more generalized solution: add a new HOOK `ekg-edit-hook`
So what I have in my mind is to add a HOOK `ekg-edit-hook` within `ekg-edit`, thus user could use it to add customizations like other -hook does. 

For my use case:
```elisp
(defun z/ekg-set-buffer-not-modified (&optional _note)
  "Set buffer state as not modified."
  (set-buffer-modified-p nil))

(defvar ekg-edit-hook nil  "Hook run after opening a note.")
(setq ekg-edit-hook '(z/ekg-set-buffer-not-modified))
(setq ekg-pre-save-hook '(z/ekg-set-buffer-not-modified))
```

Like to hear you thoughts.

**Comment by ahyatt** [comment]
I think setting the modified state to `nil` seems pretty reasonable.  After all, when you first open / capture, it really isn't modified even though we add metadata.  Perhaps it could be better, though, maybe the best thing to do is to note before changing the metadata section if the buffer is unmodified, and if it is, reset it to unmodified after changing the metadata.

Also, your use-case seems a lot like https://github.com/ahyatt/ekg/pull/19, and I have the same objection, which I don't think I've really thought of a great solution for yet.  If you have any ideas, let me know!

**Reply by qingshuizheng** [reply]
> I think setting the modified state to `nil` seems pretty reasonable.

Please add them :-)

> If you have any ideas, let me know!

Save note to ekg database from current capture/edit buffer every x seconds with a timer, unless cursor is in the metadata section. 

For ekg-capture buffer, when a capture buffer is created, tag it with a "raw" tag to make it trackable, such that if buffer is killed mistakenly, you could find it with the "raw" tag; If it is finalized normally with `C-c C-c`, remove the "raw" tag.  

This is the best I can think of, and it is what I'm using now.

**Reply by ahyatt** [reply]
Thanks!  You sent me the PR to add these yourself, so thank you!  I like the "raw" tag idea (I think "draft" might be more familiar to users, though).  That's a direction that could resolve the underlying request behind https://github.com/ahyatt/ekg/pull/19, so let me look into implementing this in ekg and see how it looks.

**Reply by ahyatt** [reply]
I've implemented this in the `develop` branch in commit 2ba885f09c8ecc06458c3f3edb232a557b4696fb.  Thank you for your suggestion!

**Reply by qingshuizheng** [reply]
Coming late. I just finished a dog-fighting with "UNIQUE constraint failed" error. Not until today I learnt how to remove all the dups with the `DB Browser for SQLite.app`. Pheeeew!!

Awesome change! The "draft" approach is a good way to review the unfinished notes. Have been using it for months, and it is doing great.

One little thing: since the `ekg-draft-tag` also works for `ekg-edit-buffer`, will it surprise users when they happen to refresh the notes buffer, and the note that are being edited disappears? Especially if the current notes buffer is "latest-modified" buffer.

**Reply by jayrajput** [reply]
I also struggled with the 'UNIQUE constraint failed' error, but thought it was something with my database. Good to know that it was not only me.

**Reply by qingshuizheng** [reply]
Nice, It was not only me. It began around 10 days ago, perhaps due to any change in the Emacs master branch? I rebuilt it around that time.

I switched to `emacsql-sqilte` for a while, but it is way slower than the builtin sqlite. Weirdly, `emacsql-sqilte` doesn't care about too much the "UNIQUE constraint", you could still query the db without firing errors.

Today I removed the dups, and back to the builtin for fast query:

step 0. BACKUP

step 1. query for dups:
```sqlite3
SELECT * FROM triples
WHERE ROWID > (
SELECT MIN(rowid) FROM triples p2  
  WHERE triples.subject = p2.subject
  AND triples.predicate = p2.predicate
  AND triples.object = p2.object
  AND triples.properties = p2.properties
);
```

step 2. remove the dups
```sqlite3
DELETE FROM triples
WHERE ROWID > (
SELECT MIN(rowid) FROM triples p2  
  WHERE triples.subject = p2.subject
  AND triples.predicate = p2.predicate
  AND triples.object = p2.object
  AND triples.properties = p2.properties
);
SELECT * FROM triples;
```

**Reply by jayrajput** [reply]
I used these steps suggested by chat GPT. It was one of those moments where I did enjoyed the power of keeping the notes in the database.

```
-- Step 1: Create a temporary table
CREATE TABLE temp_table AS
SELECT DISTINCT * FROM triples;

-- Step 2: Rename the original table
ALTER TABLE triples RENAME TO original_table;

-- Step 3: Rename the temporary table
ALTER TABLE temp_table RENAME TO triples;
```

**Reply by qingshuizheng** [reply]
Your version is neat. DISTINCT is the right method. Thanks for sharing!

**Reply by jayrajput** [reply]
I restarted my emacs multiple times and it is funny that the UNIQUE constraint strikes back again. I am definitely doing some dumb here. But need to figure it out. I do remember fixing my triples.db file.

```
triples-setup-table-for-builtin: Database error: "UNIQUE constraint failed: triples.subject, triples.predicate, triples.object, triples.properties" [2 times]
```

**Reply by qingshuizheng** [reply]
> Database error: "UNIQUE constraint failed: triples.subject, triples.predicate, triples.object, triples.properties"

Had the same error before I was able to de-dup the db. 

> perhaps due to any change in the Emacs master branch?

I think I jumped to conclusion too quickly. Will try to catch the error again under Emacs -Q. Until then, I will report back.

**Reply by ahyatt** [reply]
I've had the same issue, and it should be that being on the latest triples version (0.3.3) and running `ekg-force-upgrade`, should fix all of these problems.  I'm not sure why it didn't do this the first time you ran the upgrade, but it should leave your db with the indexes set up that ensure there's no dups.

**Reply by qingshuizheng** [reply]
@ahyatt  

I was at `develop` branch, and did not notice there was a update at `main` branch. 
Now I switch to `main` and run `ekg-force-upgrade`, looks good, thanks for noticing.

BTW, some of my notes have duplicate TITLES, how to de-dup them?

One of them:
![image](https://github.com/ahyatt/ekg/assets/78839107/7a8f5f69-16db-42fe-8611-aa5eb4e5588a)

**Reply by ahyatt** [reply]
I think the update was on the `develop` branch as well, at least I hope I didn't forget it there.

About the extra title, based on that, I'm a bit worried that you still have a problem with dups, and your db isn't set up correctly for some reason.  This worries me, so i'd like to get to the bottom of it.  First, connect (via `M-x sql-sqlite`) to your db and run 

```
SELECT * FROM triples WHERE object = 'magic/transient';
```

Let me know what you get - if you have two identical rows, then please run 
```
.indexes
```

and let me know what you get.

**Reply by qingshuizheng** [reply]
Thanks for replying.

> I think the update was on the develop branch as well

`main` branch is 1 commit ahead of `develop`.

> SELECT * FROM triples WHERE object = 'magit/transient';

This returns nothing, but this line below returns the record:
```SELECT * FROM triples WHERE object LIKE '%magit/transient%';``` (see below)

-----
Sqlite results attached:
```sql
SQLite version 3.32.3 2020-06-18 14:16:19
Enter ".help" for usage hints.

sqlite> SELECT * FROM triples WHERE object = 'magit/transient';
SELECT * FROM triples WHERE object = 'magit/transient';

sqlite>  SELECT * FROM triples WHERE object LIKE '%magit/transient%';
 SELECT * FROM triples WHERE object LIKE '%magit/transient%';
"https://github.com/magit/transient"|text/text|"- repo :: https://github.com/magit/transient
- manual :: https://magit.vc/manual/transient
- wiki :: https://github.com/magit/transient/wiki
  - https://github.com/magit/transient/wiki#third-party-examples--documentation

- demo:
  - system crafters:
    - part1: https://www.youtube.com/watch?v=VLcMgIctQBg
    - part2: https://www.youtube.com/watch?v=Q2uOWXlUpgw
    - part3: https://www.youtube.com/watch?v=GxO-xD9LoQo
    - part4: https://www.youtube.com/watch?v=rxp68l-jPSw

- artical
  - tutorial:
    - https://flonic.gitlab.io/org-blog/blog/emacs-transient-tutorial/index.html

- packages:
  - transient/magit/forge
  - org-roam(magit-section)
  - transient-showcase:
    - https://github.com/positron-solutions/transient-showcase | 5
  - transient-dwim by conao3
    - https://github.com/conao3/transient-dwim.el | 5 | macro to learn from
      - macros in either repo site or package code
  - transient-extras
    - https://github.com/haji-ali/transient-extras
  - telega/telega-transient.el
    - https://github.com/zevlg/telega.el/blob/master/contrib/telega-transient.el
      - macro: telega-transient-define-prefix-by-keymap
  - screenshot.el
    - https://github.com/tecosaur/screenshot/blob/6faa48a9/screenshot.el#L72-L87
    - https://github.com/tecosaur/screenshot/blob/master/screenshot.el
  - org-ql/org-ql-view.el:
    - https://github.com/alphapapa/org-ql/blob/master/org-ql-view.el#L696
  - gptel/gptel-transient.el
"|()
"https://github.com/magit/transient"|titled/title|"magit/transient"|(:index 0)

sqlite> .indexes
.indexes

sqlite> 
```
-----
Updated .indexes results:
```sql
sqlite> .indexes
.indexes
predicate_object_idx                   
subject_idx                            
subject_predicate_idx                  
subject_predicate_object_properties_idx
```
-----
Environment:
1. macOS Big Sur
2. Emacs -Q (emacs 30, built 2 days ago)

**Reply by ahyatt** [reply]
Hm, interesting.  The first output shouldn't have shown up as two identical titles.  I see only one actual title, and no dups.

But the lack of indexes means that dups could occur.   After you updated (I assume this was via `ekg-force-update`), it is fixed, which is great.  But it's disturbing you were missing those indexes, did you force an update before, or was this your first time?

**Reply by qingshuizheng** [reply]
> But the lack of indexes means that dups could occur.

Good to know that!

> did you force an update before, or was this your first time?

I did that several times, I don't remember when the first time was. More `ekg-force-upgrade` were done in the past 2-3 weeks.

Anyway, the db works again, and I'm gonna live with that duplicate titles.

**Reply by qingshuizheng** [reply]
Another UPDATE: 

I inspected several of my notes that have dup titles via `DB Browser for SQLite.app`, and found these `(:t t)` in properties, are they valid values?

Note 1:
![image](https://github.com/ahyatt/ekg/assets/78839107/82dd656d-6bc0-4216-baed-d44fd28e2543)

Note 2:
![image](https://github.com/ahyatt/ekg/assets/78839107/79cb54eb-da83-4b96-92d5-91c32a7ec1d1)

Thanks!

**Reply by ahyatt** [reply]
The `(:t t)` is valid, and how emacsql denotes empty property values.  I can't remember exactly the reason, but I think for some reason emacsql couldn't handle the simpler "()" value there.  I'm now realizing I should be upgrading these when you switch to sqlite, although that means if you switch back there might be dups again.

**Reply by qingshuizheng** [reply]
> The (:t t) is valid, and how emacsql denotes empty property values. 

TIL.

> that means if you switch back there might be dups again.

I won't try emacsql again, the performance is terrible. It takes a thousand year to return queries. 😆

**Comment by jayrajput** [comment]
This is a good change. I recently discovered that `s-k` in Emacs on Macos kills the buffer. I am new to using Emacs on Macos. But I did lost my drafts. Thanks for making this change. I have also remapped `s-k` to bury-buffer rather than kill-buffer.

So the draft tag be automatically applied on a new capture or edit? And only on `C-c C-c` gets removed?

**Reply by qingshuizheng** [reply]
Indeed. I almost smashed my keyboard when I hit `s-k` without save.

According to the code implemented, the "draft" tag will only be pushed to the note tags if triggered by `ekg-save-draft`, which is binded to the `save-buffer` kbd. The default is `C-x C-s`. I haven't figured out how Andrew remove the "draft" tag yet (I haven't go deep into the commit code). But the tag will eventually be removed after `C-c C-c`.

**Reply by ahyatt** [reply]
Thanks!

I only removed the draft tag when setting up the edit buffer.  So it just won't appear in the list of tags in the edit buffer, and therefore, next time you save it, it won't be saved.

**Comment by qingshuizheng** [comment]
Andrew has pushed some great changes in both `ekg` and `triples` recently. 

`ekg` now has better saving mechanism, and the dup records and dup titles issues are well taken care of. The latter see these two commits in `triples` repo: https://github.com/ahyatt/triples/commit/01f8c7c82997a4d4bebcb05b0547dd7384f557cb , https://github.com/ahyatt/triples/commit/ddeb338751fe84c1a1f1ea66be54efa7beaa15bd , respectively. 

Thanks for addressing them. @ahyatt
